rig/place_and_route/utils.py

```diff
@@ -136,3 +141,18 @@ def build_routing_tables(routes, net_keys, omit_default_routes=True):
             )
 
     return routing_tables
+
+
+class MultisourceRouteError(Exception):
+    """Indicates that two nets with the same key and mask would cause packets
+    to become duplicated.
+    """
+    def __init__(self, key, mask, coordinate):
+        self.key = key
+        self.mask = mask
+        self.x, self.y = coordinate
+
+    def __str__(self):
+        return ("Two different nets with the same key ({0.key:#010x}) and "
+                "mask ({0.mask:#010x}) fork differently at ({0.x}, {0.y})".
```

Comment:
I'll try and explain when I'm back but yes, the strict flag would be
sufficient :)
On 28 Oct 2015 10:55, "Andrew Mundy" <notifications@github.com> wrote:

> In rig/place_and_route/utils.py
> <https://github.com/project-rig/rig/pull/192#discussion_r43240788>:
>
> > @@ -136,3 +141,18 @@ def build_routing_tables(routes, net_keys, omit_default_routes=True):
> >              )
> >
> >      return routing_tables
> > +
> > +
> > +class MultisourceRouteError(Exception):
> > +    """Indicates that two nets with the same key and mask would cause packets
> > +    to become duplicated.
> > +    """
> > +    def __init__(self, key, mask, coordinate):
> > +        self.key = key
> > +        self.mask = mask
> > +        self.x, self.y = coordinate
> > +
> > +    def __str__(self):
> > +        return ("Two different nets with the same key ({0.key:#010x}) and "
> > +                "mask ({0.mask:#010x}) fork differently at ({0.x}, {0.y})".
>
> I'm still not sure I understand; do you mean that for the last new test I
> added that you'd get three routing table entries at (1, 1):
>
> (0x0, 0xf, WEST)
> (0x0, 0xf, SOUTH)
> (0x0, 0xf, SOUTH | WEST) -- This entry is the one that you'd get before
> this change
>
> The problem then is that the table ordering is important and I believe we
> currently assume that tables are indifferent to ordering.
> ------------------------------
>
> I can see an acceptable alternative being that the current behaviour is
> used if you set strict=False or something. If we do that I strongly
> believe we should at least raise a warning whenever this kind of thing
> occurs.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/project-rig/rig/pull/192/files#r43240788>.
>